src/index/src/inverted_index/create/sort/external_sort.rs

```diff
@@ -0,0 +1,439 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::{BTreeMap, VecDeque};
+use std::mem;
+use std::num::NonZeroUsize;
+use std::ops::RangeInclusive;
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use common_base::BitVec;
+use futures::stream;
+
+use crate::inverted_index::create::sort::external_provider::ExternalTempFileProvider;
+use crate::inverted_index::create::sort::intermediate_rw::{
+    IntermediateReader, IntermediateWriter,
+};
+use crate::inverted_index::create::sort::merge_stream::MergeSortedStream;
+use crate::inverted_index::create::sort::{SortOutput, SortedStream, Sorter};
+use crate::inverted_index::error::Result;
+use crate::inverted_index::{Bytes, BytesRef};
+
+/// `ExternalSorter` manages the sorting of data using both in-memory structures and external files.
+/// It dumps data to external files when the in-memory buffer crosses a certain memory threshold.
+pub struct ExternalSorter {
+    /// The index name associated with the sorting operation
+    index_name: String,
+
+    /// Manages creation and access to external temporary files
+    temp_file_provider: Arc<dyn ExternalTempFileProvider>,
+
+    /// Bitmap indicating which segments have null values
+    segment_null_bitmap: BitVec,
+
+    /// In-memory buffer to hold values and their corresponding bitmaps until memory threshold is exceeded
+    values_buffer: BTreeMap<Bytes, BitVec>,
+
+    /// Count of all rows ingested so far
+    total_row_count: usize,
+
+    /// The number of rows per group for bitmap indexing which determines how rows are
+    /// batched for indexing. It is used to determine which segment a row belongs to.
+    segment_row_count: NonZeroUsize,
+
+    /// Tracks memory usage of the buffer
+    current_memory_usage: usize,
+
+    /// The memory usage threshold at which the buffer should be dumped to an external file.
+    /// `None` indicates that the buffer should never be dumped.
+    memory_usage_threshold: Option<usize>,
+}
+
+#[async_trait]
+impl Sorter for ExternalSorter {
+    /// Pushes a value into the sorter, adding it to the in-memory buffer and dumping the buffer to
+    /// an external file if necessary
+    async fn push(&mut self, value: Option<BytesRef<'_>>) -> Result<()> {
+        self.push_n(value, 1).await
+    }
+
+    /// Pushes n identical values into the sorter, adding them to the in-memory buffer and dumping
+    /// the buffer to an external file if necessary
+    async fn push_n(&mut self, value: Option<BytesRef<'_>>, n: usize) -> Result<()> {
```

Comment:
Looks like the default implementation of `push_n` in trait is unnecessary. Maybe we want user to implement `push_n` and let `push`  be the default implementation.